### ISSUE TITLE
Synchronized Keyword

### ISSUE THREAD

**Issue by zhamri** [STIW3054-A182/Main-Issues#8]
## Instruction:

Watch the video from the link below:

https://youtu.be/RH7G-N2pa8M

1. Explain when and why we should use `Synchronized` keyword.
2. Give __ONE (1)__ example of implementing `Synchronized` keyword using Java.
3. Paste your reference here.
4. Explain what you have learned from the video besides `Synchronized` keyword.

**Comment by mimiothman** [comment]
1. The synchronized keyword can be used to ensure that only one thread at a time executes a particular section of code. This is a simple way to prevent race conditions, which occur when several threads change shared data at the same time in a way that leads to incorrect results.

2. 
 ```javascript
// A Java program to demonstrate working of 
// synchronized. 
import java.io.*; 
import java.util.*; 
  
// A Class used to send a message 
class Sender  
{ 
    public synchronized void send(String msg) 
    { 
        System.out.println("Sending\t" + msg ); 
        try 
        { 
            Thread.sleep(1000); 
        }  
        catch (Exception e)  
        { 
            System.out.println("Thread interrupted."); 
        } 
        System.out.println("\n" + msg + "Sent"); 
    } 
} 
  
// Class for send a message using Threads 
class ThreadedSend extends Thread 
{ 
    private String msg; 
    private Thread t; 
    Sender  sender; 
  
    // Recieves a message object and a string 
    // message to be sent 
    ThreadedSend(String m,  Sender obj) 
    { 
        msg = m; 
        sender = obj; 
    } 
  
    public void run() 
    { 
        // Only one thread can send a message 
        // at a time. 
        synchronized(sender) 
        { 
            // synchronizing the snd object 
            sender.send(msg); 
        } 
    } 
} 
  
// Driver class 
class SyncDemo 
{ 
    public static void main(String args[]) 
    { 
        Sender snd = new Sender(); 
        ThreadedSend S1 = 
            new ThreadedSend( " Hi " , snd ); 
        ThreadedSend S2 = 
            new ThreadedSend( " Bye " , snd ); 
  
        // Start two threads of ThreadedSend type 
        S1.start(); 
        S2.start(); 
  
        // wait for threads to end 
        try
        { 
            S1.join(); 
            S2.join(); 
        } 
        catch(Exception e) 
        { 
            System.out.println("Interrupted"); 
        } 
    } 
} 
```
3. https://www.geeksforgeeks.org/synchronized-in-java/

4. We can seen different ways of using the synchronized keyword to achieve thread synchronization. We also explored how a race condition can impact our application, and how synchronization helps us avoid that.

**Comment by lancelot9927** [comment]
## 1.
A race condition occurs when two or more threads attempt to update variable shared data at the same time. Java provides a mechanism to avoid race conditions by synchronizing thread access to shared data.
The logic marked as synchronized becomes a sync block, allowing only one thread to execute at any given time.

## 2.
```Java

public class Synchronized {

	public static void main(String[] args) throws Exception{
		add a = new add();
		
		Thread t1 = new Thread(new Runnable() {
			public void run() {
				for (int i=0;i<1000;i++) {
					a.addition();
				}
			}
		});
		
		Thread t2 = new Thread(new Runnable() {
			public void run() {
				for (int i=0;i<1000;i++) {
					a.addition();
				}
			}
		});
		t1.start();
		t2.start();
		t1.join();
		
		t2.join();
		System.out.println(a.count);
	}
}

class add{
	int count;
	 
	public synchronized void addition() {
		count=count+2;
	}
}
```
## 3.Reference
1. Guide to the Synchronized Keyword in Java.[https://www.baeldung.com/java-synchronized](https://www.baeldung.com/java-synchronized)
2. Multithreading Synchronized Keyword [https://www.youtube.com/watch?v=RH7G-N2pa8M&feature=youtu.be](https://www.youtube.com/watch?v=RH7G-N2pa8M&feature=youtu.be)

## 4.
From this video, learned about we can use Synchronized keyword to solve multi-threaded programs often come to a situation where multiple threads try to access the same resources and finally produce erroneous and unforeseen results.

**Comment by SINHUI** [comment]
## 1.
The Java **_synchronized_** keyword is an essential tool in concurrent programming in Java. Its overall purpose is to only allow one thread at a time into a particular section of code thus allowing us to protect, for example, variables or data from being corrupted by simultaneous modifications from different threads.

## 2.
```jss
class Printer{
	synchronized void printDocument(int numOfCopies, String docName) { // method 3 (Synchronized)
	   for(int i=1;i<=numOfCopies; i++) {
		   try {
			   Thread.sleep(10);
		   }catch (InterruptedException e) {
			   e.printStackTrace();
		   }
		   System.out.println(">> Printing "+ docName+" "+i);
	   }
	   
   }
}

class Liew extends Thread{
	
	Printer p1;
	Liew(Printer h1){
		p1=h1;
	}
	
	@Override
	public void run(){
		//synchronized (p1) { //---Method 2 (Synchronized) must put in both way
		p1.printDocument(10, "HuiProfile.pdf");
		//}
	}
}

class Liew2 extends Thread{
	
	Printer p1;
	Liew2(Printer h1){
		p1=h1;
	}
	@Override
	public void run(){
		//synchronized (p1) { ---Method 2 (Synchronized) must put in both way
		p1.printDocument(10, "LiewProfile.pdf");
		//}
	}
}

class Liew3 extends Thread{
	
	Printer p1;
	Liew3(Printer h1){
		p1=h1;
	}
	@Override
	public void run(){
		//synchronized (p1) { ---Method 2 (Synchronized) must put in both way
		p1.printDocument(10, "SinProfile.pdf");
		//}
	}
}
public class SyncApp {
	
	// main is representing main thread
	public static void main(String[] args) {
		
		System.out.println("==Application Started==");
		
		// We have only 1 single object of Printer
		Printer printer =new Printer();
		//printer.printDocument(10, "IshantsProfile.pdf");
		
		
		// Scenario is that we have multiple thread working on the same Printer Object 
		// If Multiple Threads are going to work on the same single Object we must Synchronize Them
		Liew h1 = new Liew(printer);
		Liew2 h2 = new Liew2(printer);
		Liew3 h3 = new Liew3(printer);
		
		h1.start();
		/*try {
			h1.join();
		} catch (InterruptedException e) {    method 1 (Synchronize)
			e.printStackTrace();
		}*/
		h3.start();
		h2.start();

		System.out.println("==Application Finished==");
	}
}
```
## 3. Reference
1. https://www.javamex.com/tutorials/synchronization_concurrency_synchronized1.shtml
2. https://www.youtube.com/watch?v=TCd8QIS-2KI

## 4. 
From this video, I have learned about the importance of **_synchronized_**, which is it can avoid multiple threads access the same method at the same time. So that **_synchronized_** will only allow one thread use the method at a time.

**Comment by lishaobin** [comment]
1. Multi-threaded programs may often come to a situation where multiple threads try to access the same resources and finally produce erroneous and unforeseen results . So it needs to be made sure by some synchronization method that only one thread can access the resource at a given point of time.
Java provides a way of creating threads and synchronizing their task by using synchronized blocks. Synchronized blocks in Java are marked with the synchronized keyword. A synchronized block in Java is synchronized on some object. All synchronized blocks synchronized on the same object can only have one thread executing inside them at a time. All other threads attempting to enter the synchronized block are blocked until the thread inside the synchronized block exits the block.

2.
```java 
import java.io.*; 
import java.util.*; 
  
// A Class used to send a message 
class Sender 
{ 
    public void send(String msg) 
    { 
        System.out.println("Sending\t"  + msg ); 
        try
        { 
            Thread.sleep(1000); 
        } 
        catch (Exception e) 
        { 
            System.out.println("Thread  interrupted."); 
        } 
        System.out.println("\n" + msg + "Sent"); 
    } 
} 
  
// Class for send a message using Threads 
class ThreadedSend extends Thread 
{ 
    private String msg; 
    private Thread t; 
    Sender  sender; 
  
    // Recieves a message object and a string 
    // message to be sent 
    ThreadedSend(String m,  Sender obj) 
    { 
        msg = m; 
        sender = obj; 
    } 
  
    public void run() 
    { 
        // Only one thread can send a message 
        // at a time. 
        synchronized(sender) 
        { 
            // synchronizing the snd object 
            sender.send(msg); 
        } 
    } 
} 
  
// Driver class 
class SyncDemo 
{ 
    public static void main(String args[]) 
    { 
        Sender snd = new Sender(); 
        ThreadedSend S1 = 
            new ThreadedSend( " Hi " , snd ); 
        ThreadedSend S2 = 
            new ThreadedSend( " Bye " , snd ); 
  
        // Start two threads of ThreadedSend type 
        S1.start(); 
        S2.start(); 
  
        // wait for threads to end 
        try
        { 
            S1.join(); 
            S2.join(); 
        } 
        catch(Exception e) 
        { 
            System.out.println("Interrupted"); 
        } 
    } 
}  
```

3.Reference 
1.[https://www.baeldung.com/java-synchronized](https://www.baeldung.com/java-synchronized)
2.[https://www.youtube.com/watch?v=RH7G-N2pa8M&feature=youtu.be](https://www.youtube.com/watch?v=RH7G-N2pa8M&feature=youtu.be)


4.From this video , i have learnend that wew can use the  synchronized to avoid multiple threads try to access the same resources and finally produce erroneous and unforeseen results ,and I also learned  different ways of using the synchronized keyword to achieve thread synchronization.

**Comment by limxinyii** [comment]
### 1.  
**when to use Synchronized keyword**
If your code is executing in a multi-threaded environment, you need synchronization for objects, which are shared among multiple threads, to avoid any corruption of state or any kind of unexpected behavior. Its overall purpose is to only allow one thread at a time into a particular section of code.
Synchronization in Java will only be needed if shared object is mutable. if your shared object is either read-only or immutable object, then you don't need synchronization, despite running multiple threads.

**Why use Synchronized keyword**
-  To prevent thread interference.
-  To prevent consistency problem.

### 2. 
```` java
//example of java synchronized method  
class Table{  
 synchronized void printTable(int n){//synchronized method  
   for(int i=1;i<=5;i++){  
     System.out.println(n*i);  
     try{  
      Thread.sleep(400);  
     }catch(Exception e){System.out.println(e);}  
   }  
  
 }  
}  
  
class MyThread1 extends Thread{  
Table t;  
MyThread1(Table t){  
this.t=t;  
}  
public void run(){  
t.printTable(5);  
}  
  
}  
class MyThread2 extends Thread{  
Table t;  
MyThread2(Table t){  
this.t=t;  
}  
public void run(){  
t.printTable(100);  
}  
}  
  
public class TestSynchronization2{  
public static void main(String args[]){  
Table obj = new Table();//only one object  
MyThread1 t1=new MyThread1(obj);  
MyThread2 t2=new MyThread2(obj);  
t1.start();  
t2.start();  
}  
}  
````
### 3.  Reference
1.  Synchronization in Java. https://www.javatpoint.com/synchronization-in-java
2. Java Synchronization Tutorial. https://javarevisited.blogspot.com/2011/04/synchronization-in-java-synchronized.html

### 4. Form this video, I have learned the use and important concept of synchronized. Since Java is a multi-threaded language where multiple threads run in parallel to complete program execution, so it is  important to using synchronized to avoid any corruption of state or any kind of unexpected behavior.

**Comment by limxinyii** [comment]
### 1.
**when to use Synchronized keyword**
If your code is executing in a multi-threaded environment, you need synchronization for objects, which are shared among multiple threads, to avoid any corruption of state or any kind of unexpected behavior. Its overall purpose is to only allow one thread at a time into a particular section of code.
Synchronization in Java will only be needed if shared object is mutable. if your shared object is either read-only or immutable object, then you don't need synchronization, despite running multiple threads.

**Why use Synchronized keyword**

- To prevent thread interference.
- To prevent consistency problem.

### 2.
````java
//example of java synchronized method  
class Table{  
 synchronized void printTable(int n){//synchronized method  
   for(int i=1;i<=5;i++){  
     System.out.println(n*i);  
     try{  
      Thread.sleep(400);  
     }catch(Exception e){System.out.println(e);}  
   }  
  
 }  
}  
  
class MyThread1 extends Thread{  
Table t;  
MyThread1(Table t){  
this.t=t;  
}  
public void run(){  
t.printTable(5);  
}  
  
}  
class MyThread2 extends Thread{  
Table t;  
MyThread2(Table t){  
this.t=t;  
}  
public void run(){  
t.printTable(100);  
}  
}  
  
public class TestSynchronization2{  
public static void main(String args[]){  
Table obj = new Table();//only one object  
MyThread1 t1=new MyThread1(obj);  
MyThread2 t2=new MyThread2(obj);  
t1.start();  
t2.start();  
}  
}  
````
### 3. Reference
1. Synchronization in Java. https://www.javatpoint.com/synchronization-in-java
2. Java Synchronization Tutorial. https://javarevisited.blogspot.com/2011/04/synchronization-in-java-synchronized.html

### 4. 
Form this video, I have learned the use and important concept of synchronized. Since Java is a multi-threaded language where multiple threads run in parallel to complete program execution, so it is important to using synchronized to avoid any corruption of state or any kind of unexpected behavior.

**Comment by lingzhongli** [comment]
1. We use the synchronized keyword when there have two Thread thread the same methods. Through the synchronized, methods can only be accessed by one thread at a time. 

2.
```js
import java.util.logging.Level;
import java.util.logging.Logger;

class First
{
 public void display(String msg)
 {
  System.out.print ("["+msg);
  try
  {
   Thread.sleep(1000);
  }
  catch(InterruptedException e)
  {
   e.printStackTrace();
  }
  System.out.println ("]");
 }
}

class Second extends Thread
{
 String msg;
 First fobj;
 Second (First fp,String str)
 {
  fobj = fp;
  msg = str;
  start();
     try {
         join();
     } catch (InterruptedException ex) {
         Logger.getLogger(Second.class.getName()).log(Level.SEVERE, null, ex);
     }
 }
 @Override
 public void run()
 {
  synchronized(fobj)       //Synchronized block
  {
   fobj.display(msg);
  }
 }
}

public class Synchronized 
{
 public static void main (String[] args)
 {
  First fnew = new First();
  Second ss = new Second(fnew, "welcome");
  Second ss1= new Second (fnew,"new");
  Second ss2 = new Second(fnew, "programmer");
 }
}
```
3. Reference
https://www.studytonight.com/java/synchronization.php
https://www.w3schools.com/java/java_modifiers.asp

4.  For the video, i also learn a method can be thread by more than one Thread at a same time. So, we need using synchronized to avoid this condition occurs.

**Comment by ChongMeiYong** [comment]
1. The **synchronized** keyword in Java provides locking, which ensures mutually exclusive access to the shared resourced resource and prevents data race. **Synchronized** also prevent reordering of code statement by the compiler which can cause a subtle concurrent issue if we don't use **synchronized** or **volatile** keyword. **Synchronization** only be needed if shred object is mutable.

2.
```javascript
public class Synchronized{  
	
	//synchronized method  
	synchronized void printTable(int n){
		for(int i=1;i<=5;i++){  
			System.out.println(n*i);  
			try{  
				Thread.sleep(400);  
			}catch(Exception e){
				System.out.println(e);
			}  
		}  
	}  
}  

public class Thread1 extends Thread{
	Synchronized s;  
	Thread1(Synchronized s){  
		this.s=s;  
	}  
	
	@Override
	public void run(){  
		s.printTable(5);  
	}
}

public class Thread2 extends Thread{
	Synchronized s;  
	Thread2(Synchronized s){  
		this.s=s;  
	}  
	
	@Override
	public void run(){  
		s.printTable(100);  
	}
}

public class TestSynchronized {

	public static void main(String[] args) {
		// TODO Auto-generated method stub
		Synchronized obj = new Synchronized();//only one object  
		Thread1 t1=new Thread1(obj);  
		Thread2 t2=new Thread2(obj);  
		t1.start();  
		t2.start();  
	}  
}
```

3. References
https://javarevisited.blogspot.com/2011/04/synchronization-in-java-synchronized.html
https://www.javatpoint.com/synchronization-in-java

4. From the video, I learned that what is the synchronied method. And also learned about multiple threads can exist within the same process and share resources such as memory, while different processes do not share these resources.

**Comment by roflinasuha** [comment]
1) Java synchronized keyword marks a block or method a critical section. A critical section is where one and only one thread is executing at a time, and the thread holds the lock for the synchronized section.
synchronized keyword helps in writing concurrent parts of the applications, to protect shared resources within this block. The synchronized keyword can be use with a code block and a method.
2) 
```js
//MathClass.java
public class MathClass
{
    synchronized void printNumbers(int n) throws InterruptedException
    {
        for (int i = 1; i <= n; i++)
        {
            System.out.println(Thread.currentThread().getName() + " :: "+  i);
            Thread.sleep(500);
        }
    }
}

//Main.java
public class Main
{
    public static void main(String args[])
    {
        final MathClass mathClass = new MathClass();
 
        //first thread
        Runnable r = new Runnable()
        {
            public void run()
            {
                try {
                    mathClass.printNumbers(3);
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
            }
        };
       
        new Thread(r, "ONE").start();
        new Thread(r, "TWO").start();
    }
}
```
3) Reference
https://howtodoinjava.com/java/multi-threading/java-synchronized/
https://www.youtube.com/watch?v=RH7G-N2pa8M&feature=youtu.be

4) In this video, i have learned that how the thread read the count in many time and to make the thread 2 do not run in the same time with thread 1 when thread 1 using a method, we can make a method  synchronized that means only one thread can work with that method. if thread 1 working with method, thread 2 have to wait. That means if you are not using synchronized at your method, it will give many threads to access that method.

**Comment by slyee96** [comment]
1. The Java synchronized keyword is an essential tool in concurrent programming in Java. Its overall purpose is to only allow one thread at a time into a particular section of code thus allowing us to protect, for example, variables or data from being corrupted by simultaneous modifications from different threads. 

2. 
package issue8;
import java.io.*; 
import java.util.*; 
class Sender{
    public void send(String msg){ 
        System.out.println("Sending\t"  + msg ); 
        try
        { 
            Thread.sleep(1000); 
        } 
        catch (Exception e) { 
            System.out.println("Thread  interrupted."); 
        } 
        System.out.println("\n" + msg + "Sent"); 
    } 
}
class ThreadedSend extends Thread { 
    private String msg; 
    private Thread t; 
    Sender  sender; 
  
    // Recieves a message object and a string 
    // message to be sent 
    ThreadedSend(String m,  Sender obj) 
    { 
        msg = m; 
        sender = obj; 
    } 
  
    public void run() 
    { 
        // Only one thread can send a message 
        // at a time. 
        synchronized(sender) 
        { 
            // synchronizing the snd object 
            sender.send(msg); 
        } 
    } 
} 
public class SyncDemo {
	public static void main(String[]args) {
		Sender snd = new Sender(); 
        ThreadedSend S1 = 
            new ThreadedSend( " Hi " , snd ); 
        ThreadedSend S2 = 
            new ThreadedSend( " Bye " , snd ); 
  
        // Start two threads of ThreadedSend type 
        S1.start(); 
        S2.start(); 
  
        // wait for threads to end 
        try
        { 
            S1.join(); 
            S2.join(); 
        } 
        catch(Exception e) 
        { 
            System.out.println("Interrupted"); 
        } 
    } 
}

3. Reference
https://www.javamex.com/tutorials/synchronization_concurrency_synchronized1.shtml
https://www.geeksforgeeks.org/synchronized-in-java/

4. In this issue,  I learned that what is the synchronized method. This synchronization is implemented in Java with a concept called monitors. Only one thread can own a monitor at a given time.

**Comment by trinanda98** [comment]
1. The Syncronized keyword is created to allow only one thread to be executed at a time. So we can control the thread independently and the resources in the thread will be easier to protect.
2.  
```java
public class MathClass
{
    void printNumbers(int n) throws InterruptedException
    {
        synchronized (this)
        {
            for (int i = 1; i <= n; i++)
            {
                System.out.println(Thread.currentThread().getName() + " :: "+  i);
                Thread.sleep(500);
            }
        }
    }
}
```

3. Reference:
-https://howtodoinjava.com/java/multi-threading/java-synchronized/
-https://www.javatpoint.com/synchronization-in-java
-https://www.studytonight.com/java/synchronization.php
-https://youtu.be/RH7G-N2pa8M

4. A single process could contain multiple threads in it and multiple threads can access the same method at the same time.

**Comment by HENGZHIYONG** [comment]
1. The synchronized keyword can be used to ensure that only one thread at a time executes a particular section of code. This is a simple way to prevent race conditions, which occur when several threads change shared data at the same time in a way that leads to incorrect results. With synchronized either entire methods or selected blocks can be single-threaded.

2.
```java
import java.io.*; 
import java.util.*; 

class Sender 
{ 
    public void send(String msg) 
    { 
        System.out.println("Sending\t"  + msg ); 
        try
        { 
            Thread.sleep(1000); 
        } 
        catch (Exception e) 
        { 
            System.out.println("Thread  interrupted."); 
        } 
        System.out.println("\n" + msg + "Sent"); 
    } 
} 

class ThreadedSend extends Thread 
{ 
    private String msg; 
    private Thread t; 
    Sender  sender; 

    ThreadedSend(String m,  Sender obj) 
    { 
        msg = m; 
        sender = obj; 
    } 
    public void run() 
    { 
        // Only one thread can send a message 
        // at a time. 
        synchronized(sender) 
        { 
            // synchronizing the snd object 
            sender.send(msg); 
        } 
    } 
} 

class SyncDemo 
{ 
    public static void main(String args[]) 
    { 
        Sender snd = new Sender(); 
        ThreadedSend S1 = 
            new ThreadedSend( " Hi " , snd ); 
        ThreadedSend S2 = 
            new ThreadedSend( " Bye " , snd ); 

        S1.start(); 
        S2.start(); 
        try
        { 
            S1.join(); 
            S2.join(); 
        } 
        catch(Exception e) 
        { 
            System.out.println("Interrupted"); 
        } 
    } 
} 
```
3.  -https://www.sitepoint.com/java-synchronized-keyword-tutorial/
    -https://www.geeksforgeeks.org/synchronized-in-java/

4.  From this video, I had learned what is Synchronized method and how important to use this method in multiple threads. By using Synchronized method hat only one thread can access the resource at a given point of time.

**Comment by Yvevonwen** [comment]
1. The synchronized keyword can be used to ensure that only one thread at a time executes a particular section of code. This is a simple way to prevent race conditions, which occur when several threads change shared data at the same time in a way that leads to incorrect results. With synchronized either entire methods or selected blocks can be single-threaded.

2.
```java
import java.io.*; 
import java.util.*; 

class Sender 
{ 
    public void send(String msg) 
    { 
        System.out.println("Sending\t"  + msg ); 
        try
        { 
            Thread.sleep(1000); 
        } 
        catch (Exception e) 
        { 
            System.out.println("Thread  interrupted."); 
        } 
        System.out.println("\n" + msg + "Sent"); 
    } 
} 

class ThreadedSend extends Thread 
{ 
    private String msg; 
    private Thread t; 
    Sender  sender; 

    ThreadedSend(String m,  Sender obj) 
    { 
        msg = m; 
        sender = obj; 
    } 
    public void run() 
    { 
        // Only one thread can send a message 
        // at a time. 
        synchronized(sender) 
        { 
            // synchronizing the snd object 
            sender.send(msg); 
        } 
    } 
} 

class SyncDemo 
{ 
    public static void main(String args[]) 
    { 
        Sender snd = new Sender(); 
        ThreadedSend S1 = 
            new ThreadedSend( " Hi " , snd ); 
        ThreadedSend S2 = 
            new ThreadedSend( " Bye " , snd ); 

        S1.start(); 
        S2.start(); 
        try
        { 
            S1.join(); 
            S2.join(); 
        } 
        catch(Exception e) 
        { 
            System.out.println("Interrupted"); 
        } 
    } 
} 
```
3.  -https://www.sitepoint.com/java-synchronized-keyword-tutorial/
    -https://www.geeksforgeeks.org/synchronized-in-java/

4.  From this video, I had learned what is Synchronized method and how important to use this method in multiple threads. By using Synchronized method hat only one thread can access the resource at a given point of time.

**Comment by AhKitt** [comment]
### 1. 
It’s used to make sure that two threads which access the same data don’t get in each others way. This could be two threads running the same method or different methods. When a method, or other block of code is synchronized, any other method/code block synchronized on the same object will wait until it is finished before taking their turn.

The keyword is used in two related ways. When used as part of a method header, the synchronization is on the object of which the method is part (so all synchronized methods of that object will always wait for each other to complete.) You can also use it on blocks of code within a method. In that case, you have to specify what object you want to synchronize on.

### 2. Example:
```java
class Table{  
 synchronized void printTable(int n){//synchronized method  
   for(int i=1;i<=5;i++){  
     System.out.println(n*i);  
     try{  
      Thread.sleep(400);  
     }catch(Exception e){System.out.println(e);}  
   }  
  
 }  
}  
  
class MyThread1 extends Thread{  
Table t;  
MyThread1(Table t){  
this.t=t;  
}  
public void run(){  
t.printTable(5);  
}  
  
}  
class MyThread2 extends Thread{  
Table t;  
MyThread2(Table t){  
this.t=t;  
}  
public void run(){  
t.printTable(100);  
}  
}  
  
public class TestSynchronization2{  
public static void main(String args[]){  
Table obj = new Table();//only one object  
MyThread1 t1=new MyThread1(obj);  
MyThread2 t2=new MyThread2(obj);  
t1.start();  
t2.start();  
}  
}  
```

### 3. Reference:

- https://www.javatpoint.com/synchronization-in-java
- https://www.quora.com/Why-is-the-synchronized-keyword-used-in-Java

### 4.
In this video, I had learned that synchronization is a very useful method in java. When there are multithreading, only one thread can access the shared resource in a time and other thread need to wait.

**Comment by alfmrza** [comment]
1. The _Synchronized_ keyword is used to make sure executing a section of a code is only done 1 thread at a time. It could be used when you want to prevent any interference from other threads.
2. Example: 
```java
public class Sysout {
    public static void print(String s) {
        System.out.println(s+"\n");
    }
}

public class TestThread extends Thread {
    String name;
    TheDemo theDemo;
    public TestThread(String name,TheDemo theDemo) {
        this.theDemo = theDemo;
        this.name = name;
        start();
    }
    @Override
    public void run() {
        theDemo.test(name);
    }
}

public class TheDemo {
    public synchronized void test(String name) {
        for(int i=0;i<10;i++) {
            Sysout.print(name + " :: "+i);
            try{
                Thread.sleep(500);
            } catch (Exception e) {
                Sysout.print(e.getMessage());
            }
        }
    }
    public static void main(String[] args) {
        TheDemo theDemo = new TheDemo();
        new TestThread("THREAD 1",theDemo);
        new TestThread("THREAD 2",theDemo);
        new TestThread("THREAD 3",theDemo);
    }
}
```
3. Reference(s): 
- https://www.javatpoint.com/synchronization-in-java
- https://www.geeksforgeeks.org/synchronized-in-java/

4. What I've learned: 
Many threads can exist within a single process and share the same resources as they do (e.g. memory).

**Comment by shyyahcheang** [comment]
**1. Explain when and why we should use Synchronized keyword.**
Multi-threaded programs may often come to a situation where multiple threads try to access the same resources and finally produce erroneous and unforeseen results. So it needs to be made sure by some synchronization method that only one thread can access the resource at a given point of time.

Java provides a way of creating threads and synchronizing their task by using synchronized blocks. Synchronized blocks in Java are marked with the synchronized keyword. A synchronized block in Java is synchronized on some object. All synchronized blocks synchronized on the same object can only have one thread executing inside them at a time. All other threads attempting to enter the synchronized block are blocked until the thread inside the synchronized block exits the block.

**2. Give ONE (1) example of implementing Synchronized keyword using Java.**
```java
public class Counter{
     
     long count = 0;
    
     public synchronized void add(long value){
       this.count += value;
     }
  }

public class CounterThread extends Thread{

     protected Counter counter = null;

     public CounterThread(Counter counter){
        this.counter = counter;
     }

     public void run() {
	for(int i=0; i<10; i++){
           counter.add(i);
        }
     }
  }

public class Example {

    public static void main(String[] args){
      Counter counter = new Counter();
      Thread  threadA = new CounterThread(counter);
      Thread  threadB = new CounterThread(counter);

      threadA.start();
      threadB.start(); 
    }
  }
```

**3. Paste your reference here.**

- https://www.geeksforgeeks.org/synchronized-in-java
- http://tutorials.jenkov.com/java-concurrency/synchronized.html

**4. Explain what you have learned from the video besides Synchronized keyword.**
Java provides the Synchronized keyword which allows us to atomize certain parts of the code. Basically, this means when there are multiple threads need to access the synchronization part of the code, there is no way that one thread will interrupt another thread. So we can be sure that everything works as expected.

**Comment by muhammadbimo1** [comment]
The synchronized keyword prevents concurrent access to a block of code or object by multiple Threads. it prevents a race condition occurs when two or more threads attempt to update mutable shared data at the same time.


```java
public class SOP {
    public static void print(String s) {
        System.out.println(s+"\n");
    }
}

public class TestThread extends Thread {
    String name;
    TheDemo theDemo;
    public TestThread(String name,TheDemo theDemo) {
        this.theDemo = theDemo;
        this.name = name;
        start();
    }
    @Override
    public void run() {
        theDemo.test(name);
    }
}

public class TheDemo {
    public synchronized void test(String name) {
        for(int i=0;i<10;i++) {
            SOP.print(name + " :: "+i);
            try{
                Thread.sleep(500);
            } catch (Exception e) {
                SOP.print(e.getMessage());
            }
        }
    }
    public static void main(String[] args) {
        TheDemo theDemo = new TheDemo();
        new TestThread("THREAD 1",theDemo);
        new TestThread("THREAD 2",theDemo);
        new TestThread("THREAD 3",theDemo);
    }
}
```
References
https://stackoverflow.com/questions/1085709/what-does-synchronized-mean
https://www.baeldung.com/java-synchronized

**Comment by prevenkumar** [comment]
**1. Explain when and why we should use Synchronized keyword.**
- Use Synchronized keyword to provide mutually exclusive access to a shared resources with multiple threads in Java. Synchronization in Java guarantees that no two threads can execute a synchronized method which requires the same lock simultaneously or concurrently. We can use Java Synchronized keyword only on synchronized method or synchronized block.
 
**2. Give ONE (1) example of implementing Synchronized keyword using Java.**
```java
//ThreadWithSynchronizationDemo.java

    class Pattern
    {

        public synchronized void DisplayA()             // Statement 1
        {
            for(int i=1;i<=50;i++)
            System.out.print("A");
        }

        public synchronized void DisplayB()             // Statement 2
        {
            for(int i=1;i<=50;i++)
            System.out.print("B");
        }

        public void DisplayC()
        {
            synchronized(this)                                      // Statement 3
            {
                for(int i=1;i<=50;i++)
                System.out.print("C");
            }
        }
    }

    class ThreadA implements Runnable
    {
        Pattern P;

        ThreadA(Pattern P)
        {
            this.P = P;
        }

        public void run()
        {
            P.DisplayA();
        }
    }

    class ThreadB implements Runnable
    {

        Pattern P;

        ThreadB(Pattern P)
        {
            this.P = P;
        }

        public void run()
        {
            P.DisplayB();
        }
    }

    class ThreadC implements Runnable
    {

        Pattern P;

        ThreadC(Pattern P)
        {
            this.P = P;
        }

        public void run()
        {
            P.DisplayC();
        }
    }

    class ThreadWithSynchronizationDemo
    {
        public static void main(String arg[])
        {

            Pattern P = new Pattern();

            ThreadA A = new ThreadA(P);
            ThreadB B = new ThreadB(P);
            ThreadC C = new ThreadC(P);


            Thread t1 = new Thread(A);
            Thread t2 = new Thread(B);
            Thread t3 = new Thread(C);

            t1.start();
            t2.start();
            t3.start();

        }
    }
```

**3.	Paste your reference here.**
https://howtodoinjava.com/java/multi-threading/java-synchronized/
https://javarevisited.blogspot.com/2011/04/synchronization-in-java-synchronized.html
https://www.tutorialspoint.com/java/java_thread_synchronization.htm
http://www.tutorialdost.com/Java-Programming-Tutorial/38-Java-Thread-Synchronization.aspx

**4. Explain what you have learned from the video besides Synchronized keyword.**
- The Synchronized keyword is makes a thread to works at a time and its makes other thread to wait 
   until the first thread finished. So, maximum one thread can work with it.

**Comment by muhamaddarulhadi** [comment]
**1. Explain when and why we should use Synchronized keyword.**
- use when sharing data between multiple invocations and to protect from data inconsistency


**2. Give ONE (1) example of implementing Synchronized keyword using Java.**
```Java
public class AccountTesting implements Runnable 
{
	private Account acct = new Account();
	public static void main(String[] args) 
	{
		AccountTesting r = new AccountTesting();
		Thread one = new Thread(r);
		Thread two = new Thread(r);
		one.setName("Ranjeet");
		two.setName("Reema");
		one.start();
		two.start();
	}
	@Override
	public void run() 
	{
		for (int x = 0; x < 5; x++) 
		{
			makeWithdrawal(10);
			if (acct.getBalance() < 0) 
			{
				System.out.println("account is overdrawn!");
			}
		}
	}
	private void makeWithdrawal(int amt) 
	{
		if (acct.getBalance() >= amt) 
		{
			System.out.println(Thread.currentThread().getName() + " is going to withdraw");
			try 
			{
				Thread.sleep(100);
			} 
			catch (InterruptedException ex) 
			{
				ex.printStackTrace();
			}
			acct.withdraw(amt);
			System.out.println(Thread.currentThread().getName() + " completes the withdrawal");
		} 
		else 
		{
			System.out.println("Not enough in account for " + Thread.currentThread().getName() + " to withdraw " + acct.getBalance());
		}
	}
}

class Account 
{
	private int balance = 50;
	public int getBalance() 
	{
		return balance;
	}
	public void withdraw(int amount) 
	{
		balance = balance - amount;
	}
}
```


**3. Paste your reference here.**

- Stackoverflow, (2013), Retrieved from https://stackoverflow.com/questions/10581366/synchronization-when-to-or-not-to-use
- GeeksforGeeks, (n.d.), Retrieved from https://www.geeksforgeeks.org/synchronized-in-java/
- w3resource, (2019), Retrieved from https://www.w3resource.com/java-tutorial/java-code-synchronization.php#EDITOR


**4. Explain what you have learned from the video besides Synchronized keyword.**

- What i learned are how to avoid the resources from inconsistency and the important of managing the thread for using the resources

**Comment by tanyujia** [comment]
**1. Explain when and why we should use Synchronized keyword.**
- it can be used when two or more threads attempt to update mutable shared data at the same time.
- because it only allow one thread at a time into a particular section of code.

**2. Give ONE (1) example of implementing Synchronized keyword using Java.**


```js

public class issue implements Runnable {

	RandomNumber a = new RandomNumber();

	public static void main(String args[]) throws InterruptedException {

		Thread t1 = new Thread(new issue());
		Thread t2 = new Thread(new issue());

		t1.start();
		t1.join();
		
		t2.start();
		t2.join();
	}

	@Override
	public void run() {
		a.printNumbers();
	}
}
```

```js
import java.util.Random;

public class RandomNumber extends Thread {

	private int number[] = new int[1000];

	public synchronized void printNumbers(){
		System.out.println("Display number");
		Random ran = new Random();
		for (int i=0;i<500;i++) {
		number[i] = ran.nextInt(900) + 100;
		System.out.println(number[i]);
	}
	}
}

```

**3. Paste your reference here.**
- https://www.javamex.com/tutorials/synchronization_concurrency_synchronized1.shtml
- https://javarevisited.blogspot.com/2011/04/synchronization-in-java-synchronized.html

**4. Explain what you have learned from the video besides Synchronized keyword.**
- From this video, I understand what is _Synchronized_ in Java, when and how it can be used. Besides that, i also know that how to solve the problem, which is "only allow one thread at a time into a particular section of code".

**Comment by mwng97** [comment]
**1. Explain when and why we should use Synchronized keyword.**

> Synchronized methods enable a simple strategy for preventing thread interference and memory consistency errors: if an object is visible to more than one thread, all reads or writes to that object's variables are done through synchronized methods.

2. **Example**

```java
package java_Synchronization;

	public class SynchDemo{
	     void printNumber(int n)
	     {  
	          synchronized(this)
	          {
	               for(int i=1;i<=5;i++)
	               {  
	                    System.out.println(n+" * "+i+" = "+n*i);  
	                    try
	                    {  
	                           Thread.sleep(400);  
	                    }
	                    catch(Exception e)
	                    {
	                           System.out.println(e);
	                    }  
	               }  
	          }  
	     } 
	}
	class MainThread1 extends Thread
	{  
	       SynchDemo t;  
	       MainThread1(SynchDemo t)
	       { 
	              this.t=t;  
	       }  
	       public void run()
	       { 
	              t.printNumber(4);  
	       }  
	}  
	class MainThread2 extends Thread
	{  
	       SynchDemo t;  
	       MainThread2(SynchDemo t)
	       {  
	              this.t=t;  
	       }
	       public void run()
	       {
	              t.printNumber(10);  
	       }
	}  
```
```java
package java_Synchronization;

public class TestSynchronizedBlock {
	  public static void main(String args[])
      {  
             SynchDemo obj = new SynchDemo();  
             MainThread1 t1=new MainThread1(obj); 
             MainThread2 t2=new MainThread2(obj);  
             t1.start();  
             t2.start();  
      } 
}

```

**3. Reference:**

> 1. https://stackoverflow.com/questions/1085709/what-does-synchronized-mean
> 2. https://www.tutorialride.com/core-java/synchronization-in-java-thread.htm
> 

4. **What have I learned?**

> From this video and the references above, i know that synchronized methods in Java can't be called in the same time from multiple threads.

**Comment by ChanJunLiang** [comment]
1.
Synchronized keyword is used to allow only have one thread executing inside them at a time. All other threads attempting to enter the synchronized block are blocked until the thread inside the synchronized block exits the block.

2.

```java
import java.io.*; 
import java.util.*; 
   
class Sender 
{ 
    public void send(String msg) 
    { 
        System.out.println("Sending\t"  + msg ); 
        try
        { 
            Thread.sleep(1000); 
        } 
        catch (Exception e) 
        { 
            System.out.println("Thread  interrupted."); 
        } 
        System.out.println("\n" + msg + "Sent"); 
    } 
} 
   
class ThreadedSend extends Thread 
{ 
    private String msg; 
    private Thread t; 
    Sender  sender; 
  
    // Recieves a message object and a string 
    // message to be sent 
    ThreadedSend(String m,  Sender obj) 
    { 
        msg = m; 
        sender = obj; 
    } 
  
    public void run() 
    { 
         
        synchronized(sender) 
        { 
            // synchronizing the snd object 
            sender.send(msg); 
        } 
    } 
} 
  
class SyncDemo 
{ 
    public static void main(String args[]) 
    { 
        Sender snd = new Sender(); 
        ThreadedSend S1 = 
            new ThreadedSend( " Hi " , snd ); 
        ThreadedSend S2 = 
            new ThreadedSend( " Bye " , snd ); 
  
        // Start two threads of ThreadedSend type 
        S1.start(); 
        S2.start(); 
  
        // wait for threads to end 
        try
        { 
            S1.join(); 
            S2.join(); 
        } 
        catch(Exception e) 
        { 
            System.out.println("Interrupted"); 
        } 
    } 
} 
```

3. reference
https://www.geeksforgeeks.org/synchronized-in-java/

4.
From this video, i've learned about how to use synchronized.

**Comment by TanShiJet** [comment]
# 1
Multi-threaded programs may often come to a situation where multiple threads try to access the same resources and finally produce erroneous and unforeseen results. So it needs to be made sure by some synchronization method that only one thread can access the resource at a given point of time.

# 2
```java
class Sender  
{ 
    public synchronized void send(String msg) 
    { 
        System.out.println("Sending\t" + msg ); 
        try 
        { 
            Thread.sleep(1000); 
        }  
        catch (Exception e)  
        { 
            System.out.println("Thread interrupted."); 
        } 
        System.out.println("\n" + msg + "Sent"); 
    } 
} 
``` 

```java
//synchronize only part of a method
class Sender  
{ 
    public void send(String msg) 
    { 
        synchronized(this) 
        { 
            System.out.println("Sending\t" + msg ); 
            try 
            { 
                Thread.sleep(1000); 
            }  
            catch (Exception e)  
            { 
                System.out.println("Thread interrupted."); 
            } 
            System.out.println("\n" + msg + "Sent"); 
        } 
    } 
} 
``` 

# 3
https://www.geeksforgeeks.org/synchronized-in-java/

# 4 
only one thread can access the resource at a given point of time by using synchronization method

**Comment by HENGZHIYONG** [comment]
1 . synchronization is used to protect access to resources that are accessed concurrently. One of the benefits of using multiple threads in an application is that each thread executes asynchronously. There are many situations in which multiple threads must share access to common objects . For example, in a database system, you might not want one thread to be updating a database record while another thread is trying to read it. In these types of cases, we need to ensure that resource will be used by only one thread at a time. Otherwise, two or more threads could access the same resource at the same time, each unaware of the other's actions. Java enables you to coordinate the actions of multiple threads using synchronized methods and synchronized statements . An object for which access is to be coordinated is accessed through the use of synchronized methods. These methods are declared with the synchronized keyword. Only one synchronized method can be invoked for an object at a given point in time. This keeps synchronized methods in multiple threads from conflicting with each other. 

2. 
```
// A Java program to demonstrate working of 
// synchronized. 
import java.io.*; 
import java.util.*; 

// A Class used to send a message 
class Sender 
{ 
	public void send(String msg) 
	{ 
		System.out.println("Sending\t" + msg ); 
		try
		{ 
			Thread.sleep(1000); 
		} 
		catch (Exception e) 
		{ 
			System.out.println("Thread interrupted."); 
		} 
		System.out.println("\n" + msg + "Sent"); 
	} 
} 

// Class for send a message using Threads 
class ThreadedSend extends Thread 
{ 
	private String msg; 
	private Thread t; 
	Sender sender; 

	// Recieves a message object and a string 
	// message to be sent 
	ThreadedSend(String m, Sender obj) 
	{ 
		msg = m; 
		sender = obj; 
	} 

	public void run() 
	{ 
		// Only one thread can send a message 
		// at a time. 
		synchronized(sender) 
		{ 
			// synchronizing the snd object 
			sender.send(msg); 
		} 
	} 
} 

// Driver class 
class SyncDemo 
{ 
	public static void main(String args[]) 
	{ 
		Sender snd = new Sender(); 
		ThreadedSend S1 = 
			new ThreadedSend( " Hi " , snd ); 
		ThreadedSend S2 = 
			new ThreadedSend( " Bye " , snd ); 

		// Start two threads of ThreadedSend type 
		S1.start(); 
		S2.start(); 

		// wait for threads to end 
		try
		{ 
			S1.join(); 
			S2.join(); 
		} 
		catch(Exception e) 
		{ 
			System.out.println("Interrupted"); 
		} 
	} 
} 
```

3.  https://www.geeksforgeeks.org/synchronized-in-java/
    http://net-informations.com/java/cjava/synchronization.htm

4.  From this video i learn how to use synchronized in java program.

**Comment by raihanwidia** [comment]
1. we use synchronized keyword when we have a multiple thread that want to use specific method alternately so it doesn't overlap the the condition of the method statement .

2.
```java

package issues8;

public class MainClass {

	public static void main(String[]args) {
		
		Println a = new Println();
		Thread task1 = new Thread(new Runnable()
				{

					@Override
					public void run() {
						// TODO Auto-generated method stub
						for(int i = 1 ; i<=1000 ; i++) {
							a.increment();
						}
					}
			
				});
		
		Thread task2 = new Thread(new Runnable()
		{

			@Override
			public void run() {
				// TODO Auto-generated method stub
				for(int i = 1 ; i<=1000 ; i++) {
					a.increment();
				}
			}
	
		});
		
		
		task1.start();
		try {
			task1.join();
			
		} catch (InterruptedException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
		task2.start();
		try {
			task2.join();
			
		} catch (InterruptedException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
		
		
		System.out.println("Count  "+ a.n);
		
		
	}

	
```
```java
package issues8;

public class Println  {

	int n;

	public synchronized  void increment() {
		n++;
	}
	
}
```
3..https://www.tutorialspoint.com/java/java_thread_synchronization.htm

4. Synchronized will be able to organized every thread that using same variable

**Comment by WongKinSin13** [comment]
### Explain when and why we should use Synchronized keyword.
When we are using more than 1 thread and 1 method is called by multiple threads, we should use Synchronized keyword so that each thread will run accordingly and completely before another thread interferes. The process is done by locking the method and only allow current thread using it to access. Once the task finishes the method is unlocked and the next requesting thread will be able to call the method. It is also a method to ensure the method are thread safe.

### Example of multiplier of 2 and 5
```java
class Multiplier{

	public synchronized void numGenerator(int n) {
		for(int i=1;i<=12;i++) {
			System.out.println("Multiplier of "+ n + ": " + i*n);
		try {
			Thread.sleep(200);
		}catch(Exception e) {System.out.print(e);}
		}
		};
	}

public class SyncTest {
	public static void main(String[]args) throws Exception{
		Multiplier c = new Multiplier();
		
		Thread t1 = new Thread(new Runnable() {
		public void run() {
				c.numGenerator(2);
		}
		});
		
		Thread t2 = new Thread(new Runnable() {
			public void run() {			
					c.numGenerator(5);
			}
			});
		
		t1.start();
		t2.start();
}
}
```
### Sample output: 
**Without synchronization**
![image](https://user-images.githubusercontent.com/37305852/55001987-b94c6080-5010-11e9-8299-3225b973984f.png)

**With synchronization**
![image](https://user-images.githubusercontent.com/37305852/55001946-a5a0fa00-5010-11e9-8086-10c33d5f9f78.png)

**Reference:** https://www.javatpoint.com/synchronization-in-java

**Feedback:** More understanding on join(), some java OOP and thread revision.

**Comment by FatihahFauzi** [comment]
1.Threads communicate primarily by sharing access to fields and the objects reference fields refer to. This form of communication is extremely efficient, but makes two kinds of errors possible: thread interference and memory consistency errors. The tool needed to prevent these errors is synchronization.
However, synchronization can introduce thread contention, which occurs when two or more threads try to access the same resource simultaneously and cause the Java runtime to execute one or more threads more slowly, or even suspend their execution. Starvation and livelock are forms of thread contention.

```javascript
2.
class Sender  
{ 
    public synchronized void send(String msg) 
    { 
        System.out.println("Sending\t" + msg ); 
        try 
        { 
            Thread.sleep(1000); 
        }  
        catch (Exception e)  
        { 
            System.out.println("Thread interrupted."); 
        } 
        System.out.println("\n" + msg + "Sent"); 
    } 
} 
```
3.
https://www.geeksforgeeks.org/synchronized-in-java/
https://docs.oracle.com/javase/tutorial/essential/concurrency/locksync.html

4.When a thread invokes a synchronized method, it automatically acquires the intrinsic lock for that method's object and releases it when the method returns. The lock release occurs even if the return was caused by an uncaught exception. Wonder what happens when a static synchronized method is invoked, since a static method is associated with a class, not an object. In this case, the thread acquires the intrinsic lock for the Class object associated with the class. Thus access to class's static fields is controlled by a lock that's distinct from the lock for any instance of the class.

**Comment by sonyhana7** [comment]
**1.** Why and when we should we use Synchronized keyword in Java is because to provide mutually exclusive access to a shared resource with multiple threads in Java. Synchronization in Java guarantees that no two threads can execute a synchronized method which requires the same lock simultaneously or concurrently.

**2.Give ONE (1) example of implementing Synchronized keyword using Java.**
   ###  MAIN CLASS

```JavaScript
public class Table {

        synchronized void printTable(int n){//synchronized method
            for(int i=1;i<=5;i++){
                System.out.println(n*i);
                try{
                    Thread.sleep(400);
                }catch(Exception e){System.out.println(e);}
            }

        }
    }

```

### #MYTHREAD1

```JavaScript
public class MyThread1  extends Thread{

    Table t;
    MyThread1(Table t){
        this.t=t;
    }
    public void run(){
        t.printTable(5);
    }

}
```
### #MYTHREAD2

```JavaScript
public class MyThread2 extends Thread{


        Table t;
        MyThread2(Table t){
            this.t=t;
        }
        public void run(){
            t.printTable(100);
        }
    }

```

### #class TestSynchronization2

```JavaScript
public` class TestSynchronization2 {

      public static void main(String args[]){

    Table obj = new Table();//only one object
    MyThread1 t1=new MyThread1(obj);
    MyThread2 t2=new MyThread2(obj);
    t1.start();
    t2.start();
}
}
```
**3.Paste your reference here.**
  a. [https://www.tutorialspoint.com/java/java_thread_synchronization.htm](url)
   b. https://www.youtube.com/watch?v=RH7G-N2pa8M&feature=youtu.be


**4. Explain what you have learned from the video besides Synchronized keyword.**

**Comment by rahimahsahar** [comment]
1. Synchronization is used to protect access to resources that are accessed concurrently. One of the benefits of using multiple threads in an application is that each thread executes asynchronously. There are many situations in which multiple threads must share access to common objects . For example, in a database system, you might not want one thread to be updating a database record while another thread is trying to read it. In these types of cases, we need to ensure that resource will be used by only one thread at a time. Otherwise, two or more threads could access the same resource at the same time, each unaware of the other's actions. Java enables you to coordinate the actions of multiple threads using synchronized methods and synchronized statements.

2. 
```java

public class MainClass {
	public class MainClass {

		public static void main(String[]args) {
			
			Println a = new Println();
			Thread task1 = new Thread(new Runnable()
					{

						@Override
						public void run() {
							// TODO Auto-generated method stub
							for(int i = 1 ; i<=1000 ; i++) {
								a.increment();
							}
						}
				
					});
			
			Thread task2 = new Thread(new Runnable()
			{

				@Override
				public void run() {
					// TODO Auto-generated method stub
					for(int i = 1 ; i<=1000 ; i++) {
						a.increment();
					}
				}
		
			});
			
			
			task1.start();
			try {
				task1.join();
				
			} catch (InterruptedException e) {
				// TODO Auto-generated catch block
				e.printStackTrace();
			}
			task2.start();
			try {
				task2.join();
				
			} catch (InterruptedException e) {
				// TODO Auto-generated catch block
				e.printStackTrace();
			}
			
			
			System.out.println("Count  "+ a.n);
			
			
		}


}
```
```java
public class Println {
	int n;

	public synchronized  void increment() {
		n++;
	}
	
}
```

3. https://www.tutorialspoint.com/java/java_thread_synchronization.htm

4. When a thread invokes a synchronized method, it automatically acquires the intrinsic lock for that method's object and releases it when the method returns. The lock release occurs even if the return was caused by an uncaught exception. Wonder what happens when a static synchronized method is invoked, since a static method is associated with a class, not an object. In this case, the thread acquires the intrinsic lock for the Class object associated with the class. Thus access to class's static fields is controlled by a lock that's distinct from the lock for any instance of the class.

**Comment by ShahifahYahya** [comment]
1. Explain when and why we should use Synchronized keyword.
The Java synchronized keyword is an essential tool in concurrent programming in Java. Its overall purpose is to only allow one thread at a time into a particular section of code thus allowing us to protect, for example, variables or data from being corrupted by simultaneous modifications from different threads. This article looks at how to use synchronized in Java to produce correctly functioning multithreaded programs. Other articles in this section look at other Java 5 concurrency facilities which have in fact superseded synchronized for certain tasks.

2.
```java

public class Issue8 {
	 void printNumber(int n)
     {  
          synchronized(this)
          {
               for(int i=1;i<=5;i++)
               {  
                    System.out.println(n+" * "+i+" = "+n*i);  
                    try
                    {  
                           Thread.sleep(400);  
                    }
                    catch(Exception e)
                    {
                           System.out.println(e);
                    }  
               }  
          }  
     } 
}
class MainThread1 extends Thread
{  
       Issue8 i;  
       MainThread1(Issue8 i)
       { 
              this.i=i;  
       }  
       public void run()
       { 
              i.printNumber(4);  
       }  
}  
class MainThread2 extends Thread
{  
       Issue8 t;  
       MainThread2(Issue8 t)
       {  
              this.t=t;  
       }
       public void run()
       {
              t.printNumber(10);  
       }
}  
```
```java
public class TestIssue8 {
	  public static void main(String args[])
      {  
             Issue8 obj = new Issue8();  
             MainThread1 t1=new MainThread1(obj); 
             MainThread2 t2=new MainThread2(obj);  
             t1.start();  
             t2.start();  
      } 
}
```
3. https://www.javamex.com/tutorials/synchronization_concurrency_synchronized1.shtml
    https://www.geeksforgeeks.org/synchronized-in-java/

4.1)Throws keyword is used to declare an exception.Exception can be handled by the calling method or it can be catches by that method.

2)It would work with this example.Because the  time taken by the machine to perform this operation is so tiny.But how can  you definitey tell how much a function will take to execute .If you wait any less than that you are running into trouble of thread collision and if u wait more..you are wasting time...And it also create hassle for the user/programmer who will call the function .
